src/plugins/vis_augmenter/public/utils/utils.ts

```diff
@@ -115,3 +129,36 @@ export const getAnyErrors = (visLayers: VisLayer[], visTitle: string): Error | u
     return undefined;
   }
 };
+
+/**
+ * Cleans up any stale saved objects caused by plugin resources being deleted. Kicks
+ * off an async call to delete the stale objs.
+ *
+ * @param augmentVisSavedObs the original augment-vis saved objs for this particular vis
+ * @param visLayers the produced VisLayers containing details if the resource has been deleted
+ * @param visualizationsLoader the visualizations saved object loader to handle deletion
+ */
+
+export const cleanupStaleObjects = (
+  augmentVisSavedObjs: ISavedAugmentVis[],
+  visLayers: VisLayer[],
+  visualizationsLoader: SavedObjectLoader
+): void => {
+  const staleVisLayers = visLayers
+    .filter((visLayer) => isVisLayerWithError(visLayer))
+    .filter(
+      (visLayerWithError) => visLayerWithError.error?.type === VisLayerErrorTypes.RESOURCE_DELETED
+    );
+  if (!isEmpty(staleVisLayers)) {
+    const objIdsToDelete = [] as string[];
+    staleVisLayers.forEach((staleVisLayer) => {
+      // Match the VisLayer to its origin saved obj to extract the to-be-deleted saved obj ID
+      const deletedPluginResourceId = staleVisLayer.pluginResource.id;
+      const savedObjId = augmentVisSavedObjs.find(
+        (savedObj) => savedObj.pluginResource.id === deletedPluginResourceId
+      )?.id;
+      if (savedObjId !== undefined) objIdsToDelete.push(savedObjId);
+    });
+    visualizationsLoader.delete(objIdsToDelete);
```

Comment:
Did you test this where this needs to do a massive cleanup and if it affects the rendering of the Dashboards page?

Reply:
No - added to something we should cover in the perf test issue #3833